pxr/imaging/hdSt/materialXFilter.cpp

```diff
@@ -690,223 +779,662 @@ _UpdatePrimvarNodes(
                 (*mxHdPrimvarMap)[texCoordName] = mxNodeDef->getType();
             }
         }
+    }
+}
 
-        // Connect the primvar node to the terminal node for HdStMaterialNetwork
-        // Create a unique name for the new connection.
-        std::string newConnName = primvarPath.GetName() + "_primvarconn";
-        HdMaterialConnection2 primvarConn;
-        primvarConn.upstreamNode = primvarPath;
-        primvarConn.upstreamOutputName = TfToken(newConnName);
-        hdNetwork->nodes[hdTerminalNodePath]
-            .inputConnections[primvarConn.upstreamOutputName] = {primvarConn};
+template <typename T>
+static bool
+_ParameterDiffersFrom(
+    HdMaterialNode2 const& terminal,
+    TfToken const& paramName,
+    T const& paramValue)
+{
+    // A connected value is always considered to differ:
+    if (terminal.inputConnections.find(paramName) != terminal.inputConnections.end()) {
+        return true;
     }
+    // Check the value itself:
+    const auto paramIt = terminal.parameters.find(paramName);
+    if (paramIt != terminal.parameters.end() && paramIt->second != paramValue)
+    {
+        return true;
+    }
+    // Assume default value is equal to paramValue.
+    return false;
 }
 
 static std::string const&
-_GetMaterialTag(HdMaterialNode2 const& terminal)
+_GetUsdPreviewSurfaceMaterialTag(HdMaterialNode2 const& terminal)
```

Comment:
Using a per nodetype function since generic code does not work all the time. Especially true for UsdPreviewSurface and gfTF since they both support `masked` mode and this prevents automatic detection from properly deducing the material tag.